proposals/0000-import-shadowing.rst

```diff
@@ -0,0 +1,311 @@
+nested ``let``-block, e.g. for the following program:
+ 
+::
+
+ module Mod (fun1, fun2) where
+
+ import A hiding (overridden)
+ import qualified B
+
+ overridden = ... importedFromA ...
+ fun1 = ... overridden ...
+ fun2 = ... B.importedFromB ... fun1 ...
+
+we can write out its explicit scoping as:
+
+::
+
+ let
+   -- imports from A
+   importedFromA = ...
+
+   -- imports from B
+   B.importedFromB = ...
+
+   -- defined in Mod
+   overridden = ... importedFromA ...
+   fun1 = ... overridden ...
+   fun2 = ... B.importedFromB ... fun1 ...
+ in
+   -- exports of Mod
+   (fun1, fun2)
+
+With our proposed scheme, the same program with ``ImportShadowing``
+turned on can be modeled as a two nested ``let`` blocks:
+
+::
+
+ let
+   -- imported from A
+   importedFromA = ...
+
+   -- imports from B
+   B.importedFromB = ...
+
+ in
+   -- defined in Mod
+   let
+     overridden = ... importedFromA ...
+     fun1 = ... overridden ...
+     fun2 = ... B.importedFromB ... fun1 ...
+   in
+     -- exports of Mod
+     (fun1, fun2)
+
+Of course, in this example, there is no observable difference between
+the two desugarings, since our module ``Mod`` was already well-scoped
+with the Haskell 2010 shadowing rules. However, if we change the
+program slightly by importing all of ``A`` wholesale:
+
+::
+
+ module Mod (fun1, fun2) where
+
+ import A
+ import qualified B
+
+ overridden = ... importedFromA ...
+ fun1 = ... overridden ...
+ fun2 = ... B.importedFromB ... fun1 ...
+
+then the desugaring using Haskell 2010 semantics leads to the
+following invalid program (note the two bindings of ``overridden`` in
+the same ``let``):
+
+::
+
+ let
+   -- imports from A
+   importedFromA = ...
+   overriden = ...
+
+   -- imports from B
+   B.importedFromB = ...
+
+   -- defined in Mod
+   overridden = ... importedFromA ...
+   fun1 = ... overridden ...
+   fun2 = ... B.importedFromB ... fun1 ...
+ in
+   -- exports of Mod
+   (fun1, fun2)
+
+Whereas the ``ImportShadowing`` version is valid:
+
+::
+
+ let
+   -- imported from A
+   importedFromA = ...
+   overridden = ...
+
+   -- imports from B
+   B.importedFromB = ...
+
+ in
+   -- defined in Mod
+   let
+     overridden = ... importedFromA ... -- This shadows the imported "overridden"!
+     fun1 = ... overridden ...
+     fun2 = ... B.importedFromB ... fun1 ...
+   in
+     -- exports of Mod
+     (fun1, fun2)
+
+Examples
+--------
+This extension shines especially when shadowing names defined in the
+``Prelude``, since hiding ``Prelude`` imports otherwise requires
+changing to an explicit import for ``Prelude``: we can go from
+
+::
+
+ module Mod where
+
+ import Prelude hiding (zip)
+
+ zip = ...
+
+to just
+
+::
+
+ module Mod where
+
+ zip = ...
+
+The above example is taken directly from `the "Import" page of the
+Haskell Wiki <https://wiki.haskell.org/Import>`_.
+   
+Effect and Interactions
+-----------------------
+Beside intra-module references, the other place where top-level
```

Comment:
This looks like an extra bit of specification, not just an effect or interaction. I included this in my little specification bullet, above.